src/util/samplebuffer.h

```diff
@@ -72,6 +74,13 @@ class SampleBuffer final {
         return m_data + offset;
     }
 
+    std::span<CSAMPLE> span() {
+        return mixxx::spanutil::spanFromPtrLen(m_data, m_size);
+    }
+    const std::span<CSAMPLE> span() const {
```

Comment:
I am unsure about this. Why does this work? I assume it also did not work above with data().
For my view "const std::span<CSAMPLE>" is a "smart pointer to the non-const CAMPLE.  
Isn't a Span() allways const. I think all sub-span functions are const anyway. 
In C++ returning an value type as const does not really make sense. 

Conclusion:  I think we can delete replace both functions and replace it with:
 `std::span<CSAMPLE> span() const {`

Reply:
By the way, this function is unused.

Reply:
> Isn't a Span() allways const.

No its not. its a view into contiguous chunk of memory. Whether the memory should be seen as const or non const is defined by the template parameter. 

>  In C++ returning an value type as const does not really make sense.

Yes, sorry that slipped through it should be this instead:

```suggestion
    std::span<const CSAMPLE> span() const {
```

> Conclusion: I think we can delete replace both functions and replace it with:
`std::span<CSAMPLE> span() const {`

No, that won't work. `std::span<T>` and `std::span<const T>` are different types. In the const member function, all class members are const, so you'd essentially try to construct mutable span into a type that is `const`. Here's a godbolt to demonstrate it: https://compiler-explorer.com/z/KaEcWxGYh

Its unfortunately a flaw of the language that you can't be generic over the constness. So we have to duplicate the function here with two different signatures in regards to constness. (Rust has the same problem [but is trying to fix that](https://blog.rust-lang.org/inside-rust/2022/07/27/keyword-generics.html))

> By the way, this function is unused.

Yes, but that doesn't matter because its intended as a library function to be used in the future. Its essentially supposed to replace `Samplebuffer::ReadableSlice`

Reply:
You demonstration is not valid for our case, because the T[100] in your test is also const, because it is part of the constant object memory. In the case of SampleBuffer only the pointer is part of the (const) object. See: 
https://compiler-explorer.com/z/9To5fvK7P

Reply:
I do not relay see the use case for `std::span<const CSAMPLE> span() const {` because SampleBuffer is more used as a factory. 
I think we will convert std::span<CSAMPLE> to std::span<const CSAMPLE> implicit on the fly via function parameters.

Reply:
Can you show me the code that achieves that? If we have a const samplebuffer, and we want a view into it. We need an implementation of `std::span<const CSAMPLE> span() const`. If you want example where we'd use that in the mixxx codebase, grep for `SampleBuffer::ReadableSlice`.

Reply:
> You demonstration is not valid for our case, because the T[100] in your test is also const, because it is part of the constant object memory. In the case of SampleBuffer only the pointer is part of the (const) object. See:

Oh wow, thats even worse then IMO. that means we have a const object that allows mutable access to the resource it manages. In Rust that's called the interior mutability pattern and its used to provide safe mutable access when otherwise prohibited. In C++ we can not enforce the safety guarantees. SampleBuffer seems to be intentionally written with the RAII guideline in mind that if you have a const Samplebuffer, you can not alter the contents of the buffer within. Your proposal explicitly breaks that if I understand correctly.

Reply:
> This means 'std::span span() const {' is valid in the same way like subspan returns a non const template type with a const function: https://en.cppreference.com/w/cpp/container/span/subspan

Thats not entirely true. while `std::span::subspan` is a const member func (which makes sense), the span it returns is templated on the `element_type`. So the `decltype(std::span<const T>:subspan())` is still `const T`. With samplebuffer its not the case, we have no way to specify that the underlying buffer is const, so we can only specify it for the SampleBuffer itself. Again, most of the samplebuffer code already explicitly makes that assumption, if the SampleBuffer is const, then the access to the underyling buffer is only const as well. So to follow that convention, we need to provide a `std::span<const CSAMPLE> span() const`.

Reply:
> SampleBuffer is more used as a factory.

Can you elaborate? IMO `SampleBuffer` is explicitly not intended as a glorified alligned_malloc. Its supposed to allocate once and then own its content. All access should either be directly through `operator[]` or the Readable-/WriteableSlice classes. `SampleBuffer::data` is an escape hatch and thus should be avoided unless absolutely necessary.

Reply:
> If we have a const samplebuffer

We never have a const samplebuffer, because it always points to non-const samples.   
https://github.com/mixxxdj/mixxx/blob/828a189b045050ca6ad5aa053e528e92ce0b8653/src/util/samplebuffer.h#L179

I have found this: 
https://github.com/mixxxdj/mixxx/blob/b0a7c9e3ce9996e87146c27ca0c766c8622134a0/src/sources/soundsourcem4a.cpp#L602

But this is another class. 

But anyway, I am in general in favor to only add functions that are used. If you are really sure we need it, we need to rename it because otherwise it makes zero difference due to the implicit conversion between `std::span<CSAMPLE>` to `std::span<const CSAMPLE>`

Reply:
> Your proposal explicitly breaks that if I understand correctly.

I think you have mixed up some things. std::span not a container it is "just" a pointer which does not own the memory it is pointed to. 
We have: 
```
const std::span<const CSAMPLE> constPointerToConstSampe; // it always points to the same buffer and can't change it 
std::span<const CSAMPLE> mutablePointerToConstSampe;  // this allows to let it point to different memory but cant change it  
const std::span<CSAMPLE> mutablePointerToConstSampe;  // it alwasy points to the same memory, which can be changed.   
std::span<const CSAMPLE> mutablePointerToConstSampe;  // this allows to let it point to different memory, which can be changed
```
Independent of this most memory functions are also const, that prevents to alter the control structure of std::span.

Reply:
> We never have a const samplebuffer, because it always points to non-const samples.


YES. But SampleBuffer is supposed to hide that! Thats the entire point of the class! And we do have loads of SampleBuffer::ReadableSlices in our codebase. And we should have more in the future. Its intended as the de-facto way of passing immutable audio buffers. 

> 
> I have found this:
> 
> https://github.com/mixxxdj/mixxx/blob/b0a7c9e3ce9996e87146c27ca0c766c8622134a0/src/sources/soundsourcem4a.cpp#L602
> 
> But this is another class.
> 

`SampleBuffer::ReadableSlice` is just a more boilerplate version of `std::span<const CSAMPLE>`. So we need a way of getting a std::span<const CSAMPLE> in our codebase. If you look at the constructor of `SampleBuffer::ReadableSlice`, it takes a `const SampleBuffer&` so its supposed to be compatible to be used with `const SampleBuffer`s. 

> But anyway, I am in general in favor to only add functions that are used. If you are really sure we need it, we need to rename it because otherwise it makes zero difference due to the implicit conversion between std::span<CSAMPLE> to std::span<const CSAMPLE>

This is not a matter of which functions are used and which aren't. Its a question of API and class design and if we don't provide these functions the API is incomplete and its usage will be clunky.

Reply:
> > Your proposal explicitly breaks that if I understand correctly.
> 
> I think you have mixed up some things. std::span not a container it is "just" a pointer which does not own the memory it is pointed to. We have:
> 
> ```
> const std::span<const CSAMPLE> constPointerToConstSampe; // it always points to the same buffer and can't change it 
> std::span<const CSAMPLE> mutablePointerToConstSampe;  // this allows to let it point to different memory but cant change it  
> const std::span<CSAMPLE> mutablePointerToConstSampe;  // it alwasy points to the same memory, which can be changed.   
> std::span<CSAMPLE> mutablePointerToConstSampe;  // this allows to let it point to different memory, which can be changed
> ```
> 
> Independent of this most memory functions are also const, that prevents to alter the control structure of std::span.

I simply don't understand what you are trying to say here.

Reply:
Ok I think I kinda wrapped my head around this. But I that still leaves major issues:
1. if we implement a `std:span<CSAMPLE> span() const` we allow mutable access to an object which has been crated by const object. While not leading to correctness issues, it goes explicitly against the entire rest of the SampleBuffer API. It breaks consistency severely. 
2. The proposed api correctness relies on the coercion with explicit type annotations. Take this code:
```cpp
const SampleBuffer a{};
const auto b = a.span();
// const everywhere, still we get mutable access just
// because of the `auto` above. Combine that with
// the former point and we have a severly confusing API
b[0] = 1;
```

Reply:
Here's another example on how the API would get confusing with your proposal: 

```cpp
const SampleBuffer a{};
// a[0] = 1; // compile error
a.span()[0] = 1; // legal
```

Reply:
> if we implement a std:span<CSAMPLE> span() const we allow mutable access to an object which has been crated by const object. 

Ah OK, that is your point. I was not aware that you are planning to use a "const SampleBuffer". For my understanding the std::span should be used to pass the samples around and the SampleBuffer serves only the purpose as a Factory and owner of the memory. So in that case I agree that the pair: 
```
std::span<CSAMPLE> span() {...}
std::span<const CSAMPLE> span() const {...}
```
Is sensible

Reply:
Well, I know that we currently don't really pass `const SampleBuffer`s around, but SampleBuffer is still a container and its written as if it were. So for simplicity's sake I'd follow that API. I agree that views into SampleBuffers should be passed around as `std::span<CSAMPLE>`, I overall still think that the two method API results in a simpler and more predictable API (which we should value IMO).

Reply:
Seems like we've found consensus. Thank you very much.